src/main/java/pl/damianszczepanik/jenkins/buildhistorymanager/BuildHistoryManager.java

```diff
@@ -40,33 +40,34 @@ public BuildHistoryManager(List<Rule> rules) {
      */
     @Override
     public void perform(Job<?, ?> job) throws IOException, InterruptedException {
+        synchronized (this) {
```

Comment:
first, this can go to method signature as you synchronize whole method

second - the problem happens when plugin is executed twice at the same time?

Reply:
second - yes this happend quite frequently on our server. 
e.g. if the hourly check and the check on start of the build is executed at the same time

Reply:
first applied